db2/client.go

```diff
@@ -0,0 +1,253 @@
+package db2
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"math/big"
+	"net/http"
+
+	"github.com/ethereum/go-ethereum/common/hexutil"
+
+	"github.com/gobitfly/eth2-beaconchain-explorer/db2/store"
+)
+
+var ErrNotFoundInCache = fmt.Errorf("cannot find hash in cache")
+var ErrMethodNotSupported = fmt.Errorf("methode not supported")
+
+type RawStoreReader interface {
+	ReadBlockByNumber(chainID uint64, number int64) (*FullBlockRawData, error)
+	ReadBlockByHash(chainID uint64, hash string) (*FullBlockRawData, error)
+	ReadBlocksByNumber(chainID uint64, start, end int64) ([]*FullBlockRawData, error)
+}
+
+type WithFallback struct {
+	roundTripper http.RoundTripper
+	fallback     http.RoundTripper
+}
+
+func NewWithFallback(roundTripper http.RoundTripper, fallback http.RoundTripper) *WithFallback {
+	return &WithFallback{
+		roundTripper: roundTripper,
+		fallback:     fallback,
+	}
+}
+
+func (r WithFallback) RoundTrip(request *http.Request) (*http.Response, error) {
+	resp, err := r.roundTripper.RoundTrip(request)
+	if err == nil {
+		// no fallback needed
+		return resp, nil
+	}
+
+	var e1 *json.SyntaxError
+	if !errors.As(err, &e1) &&
+		!errors.Is(err, ErrNotFoundInCache) &&
+		!errors.Is(err, ErrMethodNotSupported) &&
+		!errors.Is(err, store.ErrNotFound) {
+		return nil, err
+	}
+
+	return r.fallback.RoundTrip(request)
+}
+
+type BigTableEthRaw struct {
+	db      RawStoreReader
+	chainID uint64
+}
+
+func NewBigTableEthRaw(db RawStoreReader, chainID uint64) *BigTableEthRaw {
+	return &BigTableEthRaw{
+		db:      db,
+		chainID: chainID,
+	}
+}
+
+func (r *BigTableEthRaw) RoundTrip(request *http.Request) (*http.Response, error) {
+	body, err := io.ReadAll(request.Body)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		request.Body = io.NopCloser(bytes.NewBuffer(body))
+	}()
+	var messages []*jsonrpcMessage
+	var isSingle bool
+	if err := json.NewDecoder(bytes.NewReader(body)).Decode(&messages); err != nil {
+		isSingle = true
+		message := new(jsonrpcMessage)
+		if err := json.NewDecoder(bytes.NewReader(body)).Decode(message); err != nil {
+			return nil, err
+		}
+		messages = append(messages, message)
+	}
+	var resps []*jsonrpcMessage
+	for _, message := range messages {
+		resp, err := r.handle(request.Context(), message)
+		if err != nil {
+			return nil, err
+		}
+		resps = append(resps, resp)
+	}
+
+	respBody, _ := makeBody(isSingle, resps)
```

Comment:
```suggestion
	respBody, err := makeBody(len(resps) == 1, resps)
	if err != nil {
	   return nil, err
	}
```

Reply:
`len(resps) == 1` is different than isSingle.
We can receive a batch with only one request, the response need to be wrapped in an array. So len(resps) doesn't give enough informations about the caller context.